### PR TITLE
Generate Feeder File from Orafin Models

### LINKED ISSUE #677: add task to write feeder file data from an OraFin payment
The "feeder file" is space-formatted file that contains invoices and invoice lines. Each invoice data is entered in a "header data" section and each invoice line's fund distribution (could be multi-funded) and tax adjustments are entered as "line data". A new "header data" essentially means "next invoice". The end of the file has one line, "control data", that contains info about the whole file (total invoices and amount). The [Data points mapping file](https://docs.google.com/spreadsheets/d/15vSge029GYl3C9_g_izyZFbGRKt8a5j6BMBu8XETtUI/edit?usp=sharing) contains info about the different fields. The [feeder sample file](https://drive.google.com/file/d/1MQMC-h3_MQMIrrZs3jnDOboVKm-tn0sN/view?usp=sharing) is a snippet of one file. The filenaming convention for these files is currently "feederYYYYMMDD_YYMMDDHHMM".

Attempt at understanding the space-formatting of the feeder file:
Starting from 0, the first field is some 9-character length string (tbd what this is supposed to be. maybe it is "LIB"+folio invoice number?). The next section, either header or line data sections, begin at position 13. For header data, the next field (invoice number) starts at position 36, then position 76 is start of invoice date field (YYYYMMDD) with invoice amount that is left-padded with 0 (15 character length) and invoice type, then at position 131 is the terms name field. Total line length is 148 (ends at character position 147). For line data, the field containing the line type, amount, and tax code starts at position 13. The amount is left-padded with 0 and is total of 15 characters in length with the tax code appended. The field containing the external account number (fund pta) begins at position 50. The total line length for line data is 119 (ends at character position 118). The last line of the file, the control data section, is 45 in length, with the batch total amount being a left-padded with 0, 15-character length field, starting at position 29.

### LINKED ISSUE #707: calculate tax for each fund distribution when multi-funded and taxed invoice line
Per policy, acquisitions will only do split funding on percentage so that we can calculate the taxes for each fund distribution since folio doesn't give us the data. Example split on percentage: https://folio-test.stanford.edu/invoice/view/d4119478-631c-4806-b0fc-70358f596fae/line/edde167f-4c78-4764-994e-c50b5447fecd/view

algorithm:
- look at the value in fundDistribution when distributionType = "percentage" for each fundId
```
"fundDistributions" : [ {
    "code" : "ABBOTT-SUL",
    "encumbrance" : "5a0d0346-1686-420d-bda2-6a8719577da4",
    "fundId" : "829206c3-1a50-4b21-87fc-4da4139a6be7",
    "distributionType" : "percentage",
    "value" : 25.0
  }, {
    "code" : "ABERCROMB-SUL",
    "fundId" : "ac5ed83a-b683-4816-a8aa-76e8075b1405",
    "distributionType" : "percentage",
    "value" : 75.0
  } ],
```
- multiply that by the adjustmentsTotal value `"adjustmentsTotal" : 4.56,` 
- e.g. tax for fund A = `.25 * 4.56 = 1.14`, tax for fund B = `.75 * 4.56 = 3.42`
- doublecheck that tax for fund A + tax for fund B = adjustmentsTotal (?) report out if something not right?

### PR DESCRIPTION
**Pull request by jermnelson**
Fixes #677
Fixes #707 

This draft PR generates a Feeder File for one or more FOLIO invoices. Right now the PR generates close-to-correct lines for the invoices in the [samples ](https://docs.google.com/document/d/16ZYktSQ1JllF7oaRoClw4rl1CoQRlth0iKWLba9tlz0/edit) document.

TODO:
- [x] Confirm for invoice with split percentages for fund distributions
- [x] Fully test new feeder file generation code for `InvoiceLine` 

## Samples

1. https://folio-test.stanford.edu/invoice/view/fd6e5f34-101e-4dd2-8542-0fdaf7713a2b

```
LIB10592     HD668330FEEDER         15142 10592                             20221206000000001442.00DR                              IMMEDIATE       
LIB10592     DR000000000225.00USE_CA              1065032-101-KARFD-53245                                              
LIB10592     TX000000000020.54USE_CA              1065032-101-KARFD-53245                                              
LIB10592     TA-00000000020.54USE_CA                                                                                   
LIB10592     DR000000000225.00USE_CA              1065090-101-KATMX-53245                                              
LIB10592     TX000000000020.54USE_CA              1065090-101-KATMX-53245                                              
LIB10592     TA-00000000020.54USE_CA                                                                                   
LIB10592     DR000000000375.00USE_CA              1065031-111-KBAEW-53245                                              
LIB10592     TX000000000034.24USE_CA              1065031-111-KBAEW-53245                                              
LIB10592     TA-00000000034.24USE_CA                                                                                   
LIB10592     DR000000000325.00USE_CA              1065032-101-KARFD-53245                                              
LIB10592     TX000000000029.67USE_CA              1065032-101-KARFD-53245                                              
LIB10592     TA-00000000029.67USE_CA                                                                                   
LIB10592     DR000000000054.49USE_CA              1065090-101-KATMX-53245                                              
LIB10592     TX000000000025.11USE_CA              1065090-101-KATMX-53245                                              
LIB10592     TA-00000000025.11USE_CA                                                                                   
LIB10592     DR000000000245.62USE_CA              1065032-101-NAAUS-53245                                              
LIB10592     TX000000000025.11USE_CA              1065032-101-NAAUS-53245                                              
LIB10592     TA-00000000025.11USE_CA                                                                                   
LIB10592     DR000000000017.00TAX_EXEMPT          1065084-101-AALIB-55320 
```

2.  https://folio-test.stanford.edu/invoice/view/35500c7d-1d19-455e-b5e8-417e7d5d395a

```
LIB10593     HD094879FEEDER         9859 10593                              20230301000000003358.45DR                              IMMEDIATE       
LIB10593     DR000000001450.00SALES_STANDARD      1065090-101-KAUPC-53245                                              
LIB10593     TX000000000132.38SALES_STANDARD      1065090-101-KAUPC-53245                                              
LIB10593     DR000000000950.00SALES_STANDARD      1065090-101-KAUPC-53245                                              
LIB10593     TX000000000086.73SALES_STANDARD      1065090-101-KAUPC-53245                                              
LIB10593     DR000000000650.00SALES_STANDARD      1065090-101-KAUPC-53245                                              
LIB10593     TX000000000059.34SALES_STANDARD      1065090-101-KAUPC-53245                                              
LIB10593     DR000000000030.00TAX_EXEMPT          1065084-101-AALIB-55320
```

3.  https://folio-test.stanford.edu/invoice/view/583ef253-5649-46ab-a8b6-ac0df1cbd9d1

```
LIB10596     HD804584FEEDER         1033INV00084008 10596                   20221213000000076183.73DR                              N30             
LIB10596     DR000000050452.16TAX_EXEMPT          1065089-101-AALIB-53263                                              
LIB10596     DR000000020848.72TAX_EXEMPT          1065089-101-AALIB-53263                                              
LIB10596     DR000000004882.85TAX_EXEMPT          1065089-101-AALIB-53263 
```

4. https://folio-test.stanford.edu/invoice/view/1d8d3412-7595-43f6-b432-c6ca7d94ba18

```
LIB10597     HD012580FEEDER         38060 10597                             20230306-00000000024.00CR                              IMMEDIATE       
LIB10597     DR-00000000024.00USE_CA              1065036-111-AALIB-53245                                              
LIB10597     TX-00000000002.19USE_CA              1065036-111-AALIB-53245                                              
LIB10597     TA000000000002.19USE_CA  
```

5. https://folio-test.stanford.edu/invoice/view/7393322f-aeaa-4d51-84b6-b83d335f9a33
```
IB10599     HD895630FEEDER         NONUM-22S-79 10599                      20220307000000200000.00DR                              IMMEDIATE      Y
LIB10599     DR000000032000.00TAX_EXEMPT          1065031-105-KARVE-53245                                              
LIB10599     DR000000032000.00TAX_EXEMPT          1065032-101-NAAUS-53245                                              
LIB10599     DR000000018000.00TAX_EXEMPT          1065090-101-KBAEB-53245                                              
LIB10599     DR000000118000.00TAX_EXEMPT          1065090-101-AALIB-53245 
```

### CODE REVIEW

**Comment by shelleydoljack** [comment]
for this one https://folio-test.stanford.edu/invoice/view/583ef253-5649-46ab-a8b6-ac0df1cbd9d1:
```
LIB10596     HD804584FEEDER         1033INV00084008 10596                   20221213000000076183.73DR                              N30             
LIB10596     DR000000050452.16TAX_EXEMPT          1065089-101-AALIB-53263                                              
LIB10596     DR000000020848.72TAX_EXEMPT          1065089-101-AALIB-53263                                              
LIB10596     DR000000004882.85TAX_EXEMPT          1065089-101-AALIB-53263 
```
I was concerned that the folio invoice number was missing since the vendor invoice number is so long. This field is set at 40 characters since the start of the next field, invoice amount, starts at position 76. From what I see in the API docs, there is no length constraint for vendor invoice number:
```
"vendorInvoiceNo": {
      "description": "This is the number from the vendor's invoice, which is different from the folioInvoiceNo",
      "type": "string"
    }
```
The folio invoice number is not editable. It is a sequence stored in the database with these characteristics:
```
              Sequence "sul_mod_invoice_storage.invoice_number"
  Type  | Start | Minimum |       Maximum       | Increment | Cycles? | Cache 
--------+-------+---------+---------------------+-----------+---------+-------
 bigint | 10000 |       1 | 9223372036854775807 |         1 | no      |     1
 ```
So it looks like the number can get really long (that maximum value field?) Maybe we won't be using a feeder file by that time. @scanedastanford Can you confirm that our vendors have a set length for their invoice numbers? Or maybe this is a "Y2K" problem we don't really need to worry about.

**Comment by shelleydoljack** [comment]
Something isn't quite right with the calculations of the first sample invoice. Is it correct for there to be a 9.13% invoice level adjustment for the sample invoices? Is that folio calculating some "effective" tax rate or something? Otherwise, looking great so far!

File: libsys_airflow/plugins/folio/helpers/orafin_models.py [hunk 1]
```diff
@@ -94,6 +158,7 @@ def amount(self):
     def attachment_flag(self):
         if self.paymentTerms and self.paymentTerms.startswith("WILLCALL"):
```
Comment (shelleydoljack) [inline libsys_airflow/plugins/folio/helpers/orafin_models.py:159]: Do we need to make this case-insensitive? Maybe a question for Suzette.

**Comment by shelleydoljack** [comment]
> Something isn't quite right with the calculations of the first sample invoice. Is it correct for there to be a 9.13% invoice level adjustment for the sample invoices? Is that folio calculating some "effective" tax rate or something? Otherwise, looking great so far!

Oh wait. The 9.13% is folio rounding up 9.125%. Nevermind. But something is still not right about the first sample invoice.

**Comment by jermnelson** [comment]
> > Something isn't quite right with the calculations of the first sample invoice. Is it correct for there to be a 9.13% invoice level adjustment for the sample invoices? Is that folio calculating some "effective" tax rate or something? Otherwise, looking great so far!
> 
> Oh wait. The 9.13% is folio rounding up 9.125%. Nevermind. But something is still not right about the first sample invoice.

@shelleydoljack I've started creating a [test](https://github.com/sul-dlss/libsys-airflow/pull/709/files#diff-22eb6d7071932018ced7acab777557f43b74550874977fec2d7e4a3a6b099c8eR420-R465) to allow us to model these invoices that have percentage splits.

**Comment by scanedastanford** [comment]
> for this one https://folio-test.stanford.edu/invoice/view/583ef253-5649-46ab-a8b6-ac0df1cbd9d1:
> 
> ```
> LIB10596     HD804584FEEDER         1033INV00084008 10596                   20221213000000076183.73DR                              N30             
> LIB10596     DR000000050452.16TAX_EXEMPT          1065089-101-AALIB-53263                                              
> LIB10596     DR000000020848.72TAX_EXEMPT          1065089-101-AALIB-53263                                              
> LIB10596     DR000000004882.85TAX_EXEMPT          1065089-101-AALIB-53263 
> ```
> 
> I was concerned that the folio invoice number was missing since the vendor invoice number is so long. This field is set at 40 characters since the start of the next field, invoice amount, starts at position 76. From what I see in the API docs, there is no length constraint for vendor invoice number:
> 
> ```
> "vendorInvoiceNo": {
>       "description": "This is the number from the vendor's invoice, which is different from the folioInvoiceNo",
>       "type": "string"
>     }
> ```
> 
> The folio invoice number is not editable. It is a sequence stored in the database with these characteristics:
> 
> ```
>               Sequence "sul_mod_invoice_storage.invoice_number"
>   Type  | Start | Minimum |       Maximum       | Increment | Cycles? | Cache 
> --------+-------+---------+---------------------+-----------+---------+-------
>  bigint | 10000 |       1 | 9223372036854775807 |         1 | no      |     1
> ```
> 
> So it looks like the number can get really long (that maximum value field?) Maybe we won't be using a feeder file by that time. @scanedastanford Can you confirm that our vendors have a set length for their invoice numbers? Or maybe this is a "Y2K" problem we don't really need to worry about.

In our documentation, it says Symphony will only accept invoice numbers of up to 20 characters in length. So I will ask Frank what we do with invoice numbers longer than that. Oracle Financials accepts invoice numbers of up to 40 characters in length.

**Comment by scanedastanford** [comment]
> > Something isn't quite right with the calculations of the first sample invoice. Is it correct for there to be a 9.13% invoice level adjustment for the sample invoices? Is that folio calculating some "effective" tax rate or something? Otherwise, looking great so far!
> 
> Oh wait. The 9.13% is folio rounding up 9.125%. Nevermind. But something is still not right about the first sample invoice.

We were using 9.13% in Symphony because it doesn't allow 3 digit decimals, so when I re-created the invoices in FOLIO, I also used 9.13%.

**Comment by shelleydoljack** [comment]
> In our documentation, it says Symphony will only accept invoice numbers of up to 20 characters in length. So I will ask Frank what we do with invoice numbers longer than that. Oracle Financials accepts invoice numbers of up to 40 characters in length.

The total field length in the feeder file is 40 characters, so that's probably where the limit is coming from. I guess the next question is, why are we appending "<space><folio invoice number>" to the vendor invoice number if we are already putting "LIB<folio invoice number" in the internal number field? Is it for some post processing on our end or on A/P's end? Maybe we just use the vendor invoice number and not append the folio invoice number to it?

**Comment by scanedastanford** [comment]
For 1. https://folio-test.stanford.edu/invoice/view/fd6e5f34-101e-4dd2-8542-0fdaf7713a2b:
The invoice lines are out of sequence. The first one should be:
LIB10592     DR000000000375.00USE_CA              1065031-111-KBAEW-53245                                              
LIB10592     TX000000000034.24USE_CA              1065031-111-KBAEW-53245                                              
LIB10592     TA-00000000034.24USE_CA
Other than that, everything else looks great!

**Comment by scanedastanford** [comment]
> The total field length in the feeder file is 40 characters, so that's probably where the limit is coming from. I guess the next question is, why are we appending `<space><folio invoice number>` to the vendor invoice number if we are already putting `LIB<folio invoice number>` in the internal number field? Is it for some post processing on our end or on A/P's end? Maybe we just use the vendor invoice number and not append the folio invoice number to it?

I believe it's because different vendors could be using the same invoice numbers. So appending it with the Symphony/FOLIO invoice number makes it unique.

File: libsys_airflow/plugins/folio/helpers/orafin_models.py (hunk 1 above)
Comment (scanedastanford) [reply libsys_airflow/plugins/folio/helpers/orafin_models.py:160]: The field is free-text and although we will have our procedure say that it should be written in all caps, it's possible someone could make a mistake. Could it be case-insensitive and still output to the feeder correctly?

**Comment by shelleydoljack** [comment]
> For 1. https://folio-test.stanford.edu/invoice/view/fd6e5f34-101e-4dd2-8542-0fdaf7713a2b:
> The invoice lines are out of sequence. The first one should be:
> LIB10592 DR000000000375.00USE_CA 1065031-111-KBAEW-53245
> LIB10592 TX000000000034.24USE_CA 1065031-111-KBAEW-53245
> LIB10592 TA-00000000034.24USE_CA
> Other than that, everything else looks great!

@scanedastanford we can get the invoice lines in sequence (e.g. first DR is from invoice line 1, etc.), but when it comes to the split funding, shouldn't we just do the order in which it is in folio? It looks like either the delphi_payments process might have been setting an order or maybe that's the order of the split funding in Symphony. I guess we're wondering if some convoluted code needs to be created to enforce some arbitrary ordering or not.

**Comment by scanedastanford** [comment]
> > For 1. https://folio-test.stanford.edu/invoice/view/fd6e5f34-101e-4dd2-8542-0fdaf7713a2b:
> > The invoice lines are out of sequence. The first one should be:
> > LIB10592 DR000000000375.00USE_CA 1065031-111-KBAEW-53245
> > LIB10592 TX000000000034.24USE_CA 1065031-111-KBAEW-53245
> > LIB10592 TA-00000000034.24USE_CA
> > Other than that, everything else looks great!
> 
> @scanedastanford we can get the invoice lines in sequence (e.g. first DR is from invoice line 1, etc.), but when it comes to the split funding, shouldn't we just do the order in which it is in folio? It looks like either the delphi_payments process might have been setting an order or maybe that's the order of the split funding in Symphony. I guess we're wondering if some convoluted code needs to be created to enforce some arbitrary ordering or not.

Great! No, we don't need to put the funds in sequence when it is a split funding. As long as the invoice lines are in sequence, then that is enough. Thanks!

**Comment by shelleydoljack** [comment]
@scanedastanford when you say "As long as the invoice lines are in sequence", do you mean the first set of DR, TX, and/TA rows should equate to invoice line 1 and the second set of DR, TX, and TA rows should correspond to invoice line 2, etc? Or do you just mean to make sure the corresponding TX and TA rows appear directly after their DR row?

**Comment by scanedastanford** [comment]
> @scanedastanford when you say "As long as the invoice lines are in sequence", do you mean the first set of DR, TX, and/TA rows should equate to invoice line 1 and the second set of DR, TX, and TA rows should correspond to invoice line 2, etc? Or do you just mean to make sure the corresponding TX and TA rows appear directly after their DR row?

I mean that the first set of DR, TX, and/TA rows should equate to invoice line 1 and the second set of DR, TX, and TA rows should correspond to invoice line 2, etc.

**Comment by jermnelson** [comment]
> 

Thanks @scanedastanford, the current code now retains the order from the invoice lines while keeping the DR/TA/TX lines together for each fund distribution in the invoice line.

**Review by shelleydoljack — APPROVED** [review]

**Pull request merged**